arch/xtensa/core/xtensa_mmu.c

```diff
@@ -329,32 +491,143 @@ void z_xtensa_mmu_smp_init(void)
+		pf = &z_page_frames[idx];
+
+		pf->flags |= Z_PAGE_FRAME_RESERVED;
+	}
+}
+#endif /* CONFIG_ARCH_HAS_RESERVED_PAGE_FRAMES */
+
+static bool l2_page_table_map(uint32_t *l1_table, void *vaddr, uintptr_t phys,
+			      uint32_t flags, bool is_user)
 {
 	uint32_t l1_pos = (uint32_t)vaddr >> 22;
-	uint32_t pte = Z_XTENSA_PTE(phys, MMU_KERNEL_RING, flags);
 	uint32_t l2_pos = Z_XTENSA_L2_POS((uint32_t)vaddr);
 	uint32_t *table;
 
-	if (l1_page_table[l1_pos] == Z_XTENSA_MMU_ILLEGAL) {
+	sys_cache_data_invd_range((void *)&l1_table[l1_pos], sizeof(l1_table[0]));
+
+	if (is_pte_illegal(l1_table[l1_pos])) {
 		table  = alloc_l2_table();
 
 		if (table == NULL) {
 			return false;
 		}
 
-		l1_page_table[l1_pos] = Z_XTENSA_PTE((uint32_t)table, MMU_KERNEL_RING,
-				Z_XTENSA_MMU_CACHED_WT);
+		init_page_table(table, XTENSA_L2_PAGE_TABLE_ENTRIES);
+
+		l1_table[l1_pos] = Z_XTENSA_PTE((uint32_t)table, MMU_KERNEL_RING,
+						Z_XTENSA_MMU_CACHED_WT);
+
+		sys_cache_data_flush_range((void *)&l1_table[l1_pos], sizeof(l1_table[0]));
+	}
+
+	table = (uint32_t *)(l1_table[l1_pos] & Z_XTENSA_PTE_PPN_MASK);
+	table[l2_pos] = Z_XTENSA_PTE(phys, is_user ? MMU_USER_RING : MMU_KERNEL_RING, flags);
+
+	sys_cache_data_flush_range((void *)&table[l2_pos], sizeof(table[0]));
+
+	return true;
+}
+
+static inline void __arch_mem_map(void *va, uintptr_t pa, uint32_t xtensa_flags, bool is_user)
+{
+	bool ret;
+	void *vaddr, *vaddr_uc;
+	uintptr_t paddr, paddr_uc;
+	uint32_t flags, flags_uc;
+
+	if (IS_ENABLED(CONFIG_XTENSA_MMU_DOUBLE_MAP)) {
+		if (arch_xtensa_is_ptr_cached(va)) {
+			vaddr = va;
+			vaddr_uc = arch_xtensa_uncached_ptr(va);
+		} else {
+			vaddr = arch_xtensa_cached_ptr(va);
+			vaddr_uc = va;
+		}
+
+		if (arch_xtensa_is_ptr_cached((void *)pa)) {
+			paddr = pa;
+			paddr_uc = (uintptr_t)arch_xtensa_uncached_ptr((void *)pa);
+		} else {
+			paddr = (uintptr_t)arch_xtensa_cached_ptr((void *)pa);
+			paddr_uc = pa;
+		}
+
+		flags_uc = (xtensa_flags & ~Z_XTENSA_PTE_ATTR_CACHED_MASK);
+		flags = flags_uc | Z_XTENSA_MMU_CACHED_WB;
+	} else {
+		vaddr = va;
+		paddr = pa;
+		flags = xtensa_flags;
+	}
+
+	ret = l2_page_table_map(z_xtensa_kernel_ptables, (void *)vaddr, paddr,
+				flags, is_user);
+	__ASSERT(ret, "Virtual address (%p) already mapped", va);
+
+	if (IS_ENABLED(CONFIG_XTENSA_MMU_DOUBLE_MAP) && ret) {
+		ret = l2_page_table_map(z_xtensa_kernel_ptables, (void *)vaddr_uc, paddr_uc,
+					flags_uc, is_user);
+		__ASSERT(ret, "Virtual address (%p) already mapped", vaddr_uc);
 	}
 
-	table = (uint32_t *)(l1_page_table[l1_pos] & Z_XTENSA_PTE_PPN_MASK);
-	table[l2_pos] = pte;
+#ifndef CONFIG_USERSPACE
+	ARG_UNUSED(ret);
+#else
+	if (ret) {
+		sys_snode_t *node;
+		struct arch_mem_domain *domain;
+		k_spinlock_key_t key;
+
+		key = k_spin_lock(&z_mem_domain_lock);
+		SYS_SLIST_FOR_EACH_NODE(&xtensa_domain_list, node) {
+			domain = CONTAINER_OF(node, struct arch_mem_domain, node);
+
+			ret = l2_page_table_map(domain->ptables, (void *)vaddr, paddr,
```

Comment:
Is this adding the l2 page table mapping to every memory domain in the system? Why is this being done?

Reply:
yep,  all virtual-to-physical mappings are the same in all page tables. Note this is called only by `arch_mem_map` that is used by the kernel. The closes that an application gets is through z_phys_map() and k_mem_map() and any of these APIs are memory domain aware. They just map physical memory in Zephyr's virtual address.